[docs] New page for the action bar customization (#5267)
mui/mui-x: docs/data/date-pickers/custom-components/custom-components.md

---
product: date-pickers
title: Date and Time pickers - Custom components
---

# Custom components

> <p class="description">The date picker let the user select a date.</p>

Reviewer: ```suggestion
<p class="description">The date picker lets users select a date from a menu.</p>
```

## Overriding components

> The picker components allows you to override internal components with the `components` prop.

Reviewer: ```suggestion
You can override the internal elements of the component (known as "slots") using the `components` prop.
```

> If you wish to pass additional props in a component slot, you can do it using the `componentsProps` prop.

Reviewer: ```suggestion
Use the `componentsProps` prop if you need to pass additional props to a component slot.
```

> As an example, you could override the action bar and pass additional props as below.

Reviewer: ```suggestion
As an example, you could override the `ActionBar` and pass additional props to the custom component as shown below:
```

```jsx
<DatePicker
  {...otherProps}
  components={{
    ActionBar: CustomActionBar,
  }}
  componentsProps={{
    actionBar: ['clear'],
  }}
/>
```

> :::info

Reviewer: Maybe this needs to be a more severe callout?

```suggestion
:::warning
```

> The casing is different between the `components` (ActionBar) and `componentsProps` (actionBar) props.

Reviewer: To be more specific.

```suggestion
The `components` prop uses Pascal case (`ActionBar`), while `componentsProps` uses camel case (`actionBar`).
```

Author: We should probably do the same change on the data grid equivalent page when improving it then

:::

## Action bar

### Component props

> The action bar is available on all the pickers component and is located at the bottom of the picker's views.

Reviewer: ```suggestion
The action bar is available on all picker components.
It is located at the bottom of the picker's views.
```

By default, it contains no action on desktop and the action _Cancel_ and _Accept_ on mobile.

> You can override the action displayed using the `actions` prop passed to this component.

Reviewer: To be more specific and tell the reader exactly what to look for.

```suggestion
You can override the action displayed by passing the `actions` prop to the `actionBar` within `componentsProps`, as shown here:
```

```jsx
<DatePicker
  componentsProps={{
    actionBar: {
      // The actions will be the same between desktop and mobile
      actions: ['clear'],

      // The actions will be different between desktop and mobile
      actions: (variant) => (variant === 'desktop' ? [] : ['clear']),
    },
  }}
/>
```

In the example below, the action bar contains only one button to reset to today's date:

{{"demo": "ActionBarComponentProps.js"}}

#### Available actions

> The build-in `ActionBar` component supports 4 different actions:

Reviewer: ```suggestion
The built-in `ActionBar` component supports four different actions:
```
